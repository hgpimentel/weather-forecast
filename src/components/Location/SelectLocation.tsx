import React, { useContext } from "react";
import styled from "styled-components";
import CenteredContainer from "../Container/CenteredContainer";
import ColumnContainer from "../Container/ColumnContainer";
import LocationContext, { Location } from "../../store/location-context";

const CustomCenteredContainer = styled(CenteredContainer)`
  align-items: center;
  justify-content: space-evenly;
  padding: 1rem;
  border-bottom: 2px solid teal;

  & > * {
    flex-grow: 1;
    max-width: 100px;
    display: flex;
    justify-content: center;
  }

  button {
    width: 75px;
    text-align: center;
    margin-top: 0.5rem;
    background-color: teal;
    color: white;
    display: inline-block;
    border-radius: 5px;
    padding: 5px;
    border: none;
    cursor: pointer;

    &:hover {
      background-color: rgba(0, 128, 128, 0.75);
    }
  }
`;

const SelectLocation: React.FC<{ closeModal: () => void }> = ({
  closeModal,
}) => {
  const lsItem = localStorage.getItem("locations");
  const locations = lsItem ? (JSON.parse(lsItem) as Location[]) : null;
  const { onLocationChange } = useContext(LocationContext);

  const onSelect = (location: Location) => {
    onLocationChange(location);
    closeModal();
  };

  return (
    <ColumnContainer>
      <div>
        <h3>Saved locations</h3>
      </div>
      {locations ? (
        <>
          {locations.map((location, index) => {
            return (
              <CustomCenteredContainer key={index}>
                <div>{location.city}</div>
                <div>{location.country}</div>
                <div>{location.lat}</div>
                <div>{location.long}</div>
                <div>
                  <button onClick={() => onSelect(locations[index])}>
                    Select
                  </button>
                </div>
              </CustomCenteredContainer>
            );
          })}
        </>
      ) : (
        <div>
          <h3>No locations found!</h3>
        </div>
      )}
    </ColumnContainer>
  );
};

export default SelectLocation;
