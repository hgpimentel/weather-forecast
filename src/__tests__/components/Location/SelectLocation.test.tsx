import renderer from "react-test-renderer";
import SelectLocation from "../../../components/Location/SelectLocation";
import { render, screen } from "@testing-library/react";

describe("Select location component", () => {
  it("renders correctly", () => {
    const component = renderer
      .create(<SelectLocation closeModal={jest.fn()} />)
      .toJSON();

    expect(component).toMatchSnapshot();
  });

  it("has elements", () => {
    const localStorageMock = (function () {
      var store: any = {
        locations:
          '[{"city":"Porto","country":"Portugal","lat":41.1496,"long":-8.611}]',
      };

      return {
        getItem: function (key: string) {
          return store[key] || null;
        },
      };
    })();

    Object.defineProperty(window, "localStorage", {
      value: localStorageMock,
    });

    render(<SelectLocation closeModal={jest.fn()} />);

    expect(screen.getByRole("heading")).toHaveTextContent("Saved locations");
  });
});
