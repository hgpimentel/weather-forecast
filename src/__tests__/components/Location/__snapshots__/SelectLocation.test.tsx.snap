// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select location component renders correctly 1`] = `null`;
